### PR TITLE
feat(greeks): Garman-Kohlhagen FX Greeks (delta/gamma/vega/theta/rho_d/rho_f)

### LINKED ISSUE #401: Add Greeks for Garman-Kohlhagen FX pricing model (pricing/greeks)
## Overview

Implement Greeks (Delta, Gamma, Vega, Theta, and two Rhos) for the Garman-Kohlhagen FX pricing model. Garman-Kohlhagen is the standard for options on foreign exchange. FX Greeks differ from equity Greeks in that they depend on two interest rates (domestic and foreign) and have two separate Rho sensitivities.

## Context

- Related PRs: #399 (Garman-Kohlhagen pricing kernel)
- Model reference: Garman & Kohlhagen (1983) "Foreign currency option values"
- Existing Greeks framework: `src/greeks/` with Black-Scholes, Monte-Carlo Greeks
- Field mapping: `underlying_price` → spot FX S, `risk_free_rate` → r_d (domestic), `dividend_yield` → r_f (foreign, since GK ≡ BSM with q=r_f)
- Structural note: GK Greeks are **mathematically equivalent** to BSM Greeks post-transformation, but FX conventions use spot-delta (not forward-delta) and split rho into `rho_d` (domestic) and `rho_f` (foreign)

## Formulas

All Greeks use spot S and two interest rates.

**Call delta (spot)**: `Δ_call = e^(-r_f·T) · N(d1)`
**Put delta (spot)**: `Δ_put = -e^(-r_f·T) · N(-d1)`

**Gamma** (same for calls/puts): `Γ = e^(-r_f·T) · n(d1) / (S · σ · √T)` where n = std normal pdf

**Vega** (same for calls/puts): `ν = S · e^(-r_f·T) · n(d1) · √T` (per 1% vol change)

**Theta** (per calendar day): Complex formula including both r_d and r_f terms (see Garman & Kohlhagen 1983 or Hull)

**Rho domestic** (per 1% domestic rate change): `ρ_d = K · T · e^(-r_d·T) · N(d2)` for calls, `= -K · T · e^(-r_d·T) · N(-d2)` for puts

**Rho foreign** (per 1% foreign rate change): `ρ_f = -S · T · e^(-r_f·T) · N(d1)` for calls, `= S · T · e^(-r_f·T) · N(-d1)` for puts

## Tasks

- [ ] `src/greeks/garman_kohlhagen.rs` — delta, gamma, vega, theta, rho_domestic, rho_foreign functions (call/put variants)
- [ ] Helper `calculate_d_values_garman_kohlhagen` if needed (verify deduplication vs pricing module)
- [ ] Trait `GarmanKohlhagenGreeks` with default implementations (pattern from `BlackScholesGreeks`)
- [ ] Tests: reference values (Garman & Kohlhagen paper, Hull, Wystup), Greek identities (delta parity), monotonicity in S/r_d/r_f, edge cases
- [ ] Example: `examples/examples_pricing/src/bin/garman_kohlhagen_greeks.rs` — compute Greeks for USD/EUR FX option scenario
- [ ] Docs: header in `src/greeks/mod.rs`, inline `///` per function, explain r_d/r_f split
- [ ] Re-exports: `pub use garman_kohlhagen::{GarmanKohlhagenGreeks, ...}` in `src/greeks/mod.rs`

## Acceptance criteria

- [ ] Delta monotonicity: call delta 0..e^(-r_f·T), put delta between -e^(-r_f·T)..0, delta parity
- [ ] Gamma > 0 always (convexity)
- [ ] Vega > 0 always
- [ ] Rho sensitivities have correct sign (positive for long calls in r_d, negative in r_f; opposite for puts)
- [ ] Hull and Garman & Kohlhagen reference values to 1e-3 tolerance (2-3 FX examples)
- [ ] Cross-check vs Black-Scholes Greeks with S and q=r_f to 1e-9 (delta, gamma, vega post-transformation should match exactly)
- [ ] Zero volatility → error
- [ ] Unsupported option types (American, exotic) → appropriate error
- [ ] Theta formula tested against numerical differentiation (Black-Scholes theta benchmark adapted for GK)
- [ ] `cargo clippy --all-targets --all-features --workspace -- -D warnings` clean
- [ ] `cargo fmt --all --check` clean
- [ ] `cargo test --all-features --workspace` clean
- [ ] `cargo build --release` clean
- [ ] All `pub` items documented

## References

- Garman, M. B., & Kohlhagen, S. W. (1983). "Foreign currency option values." Journal of International Money and Finance, 2(3), 231–237.
- Hull, J. (2017). *Options, Futures, and Other Derivatives* (10th ed.). Prentice Hall. [Chapter 17 on FX options]
- Wystup, U. (2006). *FX Options and Structured Products*. Wiley.

### PR DESCRIPTION
**Pull request by joaquinbejar**
## Summary

Closed-form Greeks for the **Garman–Kohlhagen (1983)** FX option pricing model, completing the FX story started in #399. Closes #401.

GK prices European options on a foreign-exchange spot `S` (domestic per unit of foreign), with the foreign currency earning interest at rate `r_f`. Structurally GK ≡ BSM with `q = r_f`, but FX options carry two rate sensitivities (domestic and foreign) instead of one.

## Public surface

New module `src/greeks/garman_kohlhagen.rs`:

- `pub fn delta_gk(option: &Options) -> Result<Decimal, GreeksError>`
- `pub fn gamma_gk(option: &Options) -> Result<Decimal, GreeksError>`
- `pub fn vega_gk(option: &Options)  -> Result<Decimal, GreeksError>` — per 1 % vol
- `pub fn theta_gk(option: &Options) -> Result<Decimal, GreeksError>` — per calendar day
- `pub fn rho_domestic_gk(option: &Options) -> Result<Decimal, GreeksError>` — per 1 % `r_d`
- `pub fn rho_foreign_gk(option: &Options)  -> Result<Decimal, GreeksError>` — per 1 % `r_f`
- `pub trait GarmanKohlhagenGreeks` mirroring the existing `GarmanKohlhagen` pricing trait

FX field mapping: `risk_free_rate → r_d`, `dividend_yield → r_f`, `underlying_price → S`. Greek units mirror the BSM module: vega ÷ 100, theta ÷ 365, both rhos ÷ 100. Quantity scales linearly; `Side::Short` flips the delta sign only (consistent with `greeks::delta`).

## Formulas

```
d1 = [ln(S/K) + (r_d − r_f + σ²/2)·T] / (σ·√T)
d2 = d1 − σ·√T

Δ_call = e^(-r_f·T) · N(d1)
Δ_put  = -e^(-r_f·T) · N(-d1)

Γ      = e^(-r_f·T) · n(d1) / (S · σ · √T)

ν      = S · e^(-r_f·T) · n(d1) · √T

Θ_call = -S·e^(-r_f T)·n(d1)·σ/(2√T) − r_d·K·e^(-r_d T)·N(d2)  + r_f·S·e^(-r_f T)·N(d1)
Θ_put  = -S·e^(-r_f T)·n(d1)·σ/(2√T) + r_d·K·e^(-r_d T)·N(-d2) − r_f·S·e^(-r_f T)·N(-d1)

ρ_d_call =  K · T · e^(-r_d·T) · N(d2)
ρ_d_put  = -K · T · e^(-r_d·T) · N(-d2)

ρ_f_call = -S · T · e^(-r_f·T) · N(d1)
ρ_f_put  =  S · T · e^(-r_f·T) · N(-d1)
```

## Note on BSM Greeks

This module implements the carry-adjusted Garman–Kohlhagen formulas directly using the `b = r_d − r_f` cost-of-carry term in `d1`/`d2`. It does **not** delegate to the existing BSM Greeks (`crate::greeks::delta`/`gamma`/...) because those compute `d1` from `risk_free_rate` only and then multiply the result by `e^(-qT)` — a mismatch between the d-values and the discount factor that yields incorrect numbers when `dividend_yield ≠ 0`. The pricing kernels are unaffected (they go through `calculate_d_values`, which does include `−q` in the drift). A separate issue should be filed to fix the BSM Greeks; for GK we needed correct FX numbers, hence the standalone implementation.

## Tests (18, all passing)

`#[cfg(test)] mod tests` in `src/greeks/garman_kohlhagen.rs`:

- **Range**:
  - `Δ_call ∈ (0, e^(-r_f·T))`, `Δ_put ∈ (-e^(-r_f·T), 0)`
- **Spot delta-parity**: `Δ_call − Δ_put = e^(-r_f·T)` to 1e-9
- **Positivity / symmetry**:
  - `Γ > 0`, `ν > 0` across moneyness for both call and put
  - `Γ_call = Γ_put`, `ν_call = ν_put` (Black–Scholes invariant)
- **Rho signs**: long call → +ρ_d, -ρ_f; long put → -ρ_d, +ρ_f
- **BSM equivalence at `q = 0`** (bit-exact) — the only case where the buggy BSM Greeks agree
- **Theta vs numerical price difference** (1-day bump, 5e-3 tolerance)
- **FX call-delta reference** (S=0.98, K=1.00, r_d=5 %, r_f=4 %, T=4/12, σ=10 %) → Δ ≈ 0.3909 to 1e-3
- **Error paths**:
  - zero volatility on every Greek → `GreeksError`
  - American / Bermuda / exotic → `GreeksError::Pricing(UnsupportedOptionType)`
- Trait `GarmanKohlhagenGreeks` round-trips against the free functions
- `Side::Short` negates delta
- Quantity scales linearly

## Example

`examples/examples_pricing/src/bin/garman_kohlhagen_greeks.rs` — 6-month EUR/USD (σ=10 %, r_d=4.5 %, r_f=2.5 %), walks ATM / ITM / OTM calls and puts, prints all Greeks, demonstrates `Δ_call − Δ_put = e^(-r_f·T)` (zero residual) and trait usage on a wrapper type.

## Acceptance criteria (from #401)

- [x] Delta monotonicity (call ∈ (0, e^(-r_f·T)), put ∈ (-e^(-r_f·T), 0)) and parity
- [x] Gamma > 0
- [x] Vega > 0
- [x] Rho signs (long call: +ρ_d, -ρ_f; long put: -ρ_d, +ρ_f)
- [x] Hull / Wystup-style FX reference values to 1e-3
- [x] Cross-check vs BSM Greeks at `q = 0` (bit-exact). Note: the issue's "1e-9 cross-check at `q = r_f`" can only hold once the existing BSM Greeks are corrected; the discrepancy is documented above and limited to that one criterion.
- [x] Zero volatility → error
- [x] American / Bermuda / exotic → `UnsupportedOptionType`
- [x] Theta vs numerical differentiation
- [x] `cargo clippy --all-targets --all-features --workspace -- -D warnings` clean
- [x] `cargo fmt --all --check` clean
- [x] `cargo test --lib --all-features` clean (3819 passed)
- [x] `cargo build --release` clean
- [x] All `pub` items documented

## Test plan

- [x] `cargo test --lib --all-features` — 3819 passed
- [x] `cargo clippy --all-targets --all-features --workspace -- -D warnings` — clean
- [x] `cargo fmt --all --check` — clean
- [x] `cargo build --release` — clean
- [x] `cargo run --bin garman_kohlhagen_greeks -p examples_pricing` — identity holds to zero
- [ ] CI green

### CODE REVIEW

File: src/greeks/garman_kohlhagen.rs [hunk 1]
```diff
@@ -0,0 +1,769 @@
...
+    )?;
+    let d2_value = d2(
+        option.underlying_price,
+        option.strike_price,
+        b,
+        years,
+        option.implied_volatility,
+    )?;
+    Ok((d1_value, d2_value))
+}
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/greeks/garman_kohlhagen.rs:152]: Good catch. Added a `time_to_expiry` helper that mirrors the BSM `T==0` short-circuit: delta returns the discrete intrinsic-state value (±1 ITM, 0 OTM, side-flipped) and gamma/vega/theta/ρ_d/ρ_f return zero. Five new tests cover it.

File: src/greeks/garman_kohlhagen.rs [hunk 2]
```diff
@@ -0,0 +1,769 @@
...
+            let minus = d_mul(r_d_k_df_d, big_n(d2_v)?, "greeks::gk::theta::call::minus")?;
+            let plus = d_mul(r_f_s_df_f, big_n(d1_v)?, "greeks::gk::theta::call::plus")?;
+            d_sub(common + plus, minus, "greeks::gk::theta::call::sum")?
+        }
+        OptionStyle::Put => {
+            // Θ_put = common + r_d·K·e^(-r_d T)·N(-d2) − r_f·S·e^(-r_f T)·N(-d1)
+            let plus = d_mul(r_d_k_df_d, big_n(-d2_v)?, "greeks::gk::theta::put::plus")?;
+            let minus = d_mul(r_f_s_df_f, big_n(-d1_v)?, "greeks::gk::theta::put::minus")?;
+            d_sub(common + plus, minus, "greeks::gk::theta::put::sum")?
+        }
```
[the remark this answers is not in the document]
Comment (joaquinbejar) [reply src/greeks/garman_kohlhagen.rs:349]: Done, swapped both `common + plus` calls in `theta_gk` for `d_add` so the whole path goes through the checked helpers in `src/model/decimal.rs`.

**Comment by joaquinbejar** [comment]
Thanks for the review. Both inline comments addressed in d5b7c2e8 — added BSM-style `T==0` short-circuiting and switched the theta path to `d_add` for consistent checked arithmetic.

**Comment by joaquinbejar** [comment]
Thanks for catching those edge cases. Fixed both the T=0 handling (now mirrors BSM behavior) and the theta calculation to use checked arithmetic. All tests pass including the new T=0 scenarios.
